contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
+            (uint256 newLoose, ) = liquidatePosition(amountToFree.sub(wantBal));
+
+            // if we didnt free enough money, prioritize paying down debt before taking profit
+            if (newLoose < amountToFree) {
+                if (newLoose <= _debtPayment) {
+                    _profit = 0;
+                    _loss += _debtPayment.sub(newLoose);
+                    _debtPayment = newLoose;
+                } else {
+                    _profit = newLoose.sub(_debtPayment);
+                }
+            }
+        }
     }
 
     function adjustPosition(uint256 _debtOutstanding) internal override {
-        // TODO: Do something to invest excess `want` tokens (from the Vault) into your positions
-        // NOTE: Try to adjust positions so that `_debtOutstanding` can be freed up on *next* harvest (not immediately)
+        if (emergencyExit) {
+            return;
+        }
+
+        uint256 wantBalance = balanceOfWant();
+
+        uint256 shares = 0;
+
+        if (wantBalance > dustThreshold) {
+            (, shares) = depositInBento(wantBalance);
+        }
+
+        uint256 sharesInBento = sharesInBento();
+
+        if (sharesInBento > wantToBentoShares(dustThreshold, false)) {
+            // Get highest interest rate pair
+            uint256 highestInterestIndex =
+                highestInterestPairIndex(sharesInBento);
+
+            depositInKashiPair(highestInterestIndex, sharesInBento);
+        }
     }
 
     function liquidatePosition(uint256 _amountNeeded)
         internal
         override
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
+        uint256 wantBalance = balanceOfWant();
+        if (_amountNeeded > wantBalance) {
+            uint256 amountToFree = _amountNeeded.sub(wantBalance);
+
+            uint256 deposited = estimatedTotalAssets().sub(wantBalance);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (amountToFree > 0) {
+                uint256 sharesNeeded = wantToBentoShares(amountToFree, true);
+                uint256 bentoShares = sharesInBento();
+
+                uint256 sharesToFreeFromKashi =
+                    bentoShares <= sharesNeeded
+                        ? sharesNeeded.sub(bentoShares)
+                        : 0;
+
+                uint256 sharesFreedFromKashi = 0;
+
+                // Find the lowest apr pair with at least the lesser of
+                //   - the amount to free
+                //   - the mean assets per pair
+                uint256 lowestInterestIndex =
+                    lowestInterestPairIndex(
+                        Math.min(
+                            sharesToFreeFromKashi,
```

Comment:
if bentoShares are less than sharesNeeded this is 0, is this correct?

Reply:
I'll add a conditional to prevent evaluating this branch.